analyses/cell-type-ewings/exploratory_analysis/05-cluster-exploration.Rmd

```diff
@@ -0,0 +1,626 @@
+gene_exp_df <- sce_list |>
+  purrr::map(\(sce) {
+    cell_types |>
+      purrr::map(\(type){
+        calculate_sum_markers(marker_genes_df, sce, type)
+      }) |>
+      purrr::reduce(dplyr::inner_join, by = "barcodes")
+  }) |>
+  dplyr::bind_rows(.id = "library_id")
+
+# join sum expression columns with classification df
+classification_df <- classification_df |>
+  dplyr::left_join(gene_exp_df, by = c("barcodes", "library_id"))
+```
+
+
+## Clustering
+
+Below we perform Louvain, Jaccard clustering, varying `k`. 
+The minimum `k` is 5 and the maximum `k` is 40 with a step size of 5. 
+
+```{r}
+# perform clustering across k = 5-40 with increments of 5
+all_cluster_results <- sce_list |>
+  purrr::map(cluster_sweep) |>
+  dplyr::bind_rows(.id = "library_id")
+```
+
+```{r}
+# get umap embeddings and combine into a data frame with cluster assignments
+umap_df <- sce_list |>
+  purrr::map(\(sce){
+    df <- sce |>
+      scuttle::makePerCellDF(use.dimred = "UMAP") |>
+      # replace UMAP.1 with UMAP1 and get rid of excess columns
+      dplyr::select(barcodes, UMAP1 = UMAP.1, UMAP2 = UMAP.2)
+  }) |>
+  dplyr::bind_rows(.id = "library_id") |>
+  dplyr::left_join(all_cluster_results, by = c("library_id", "barcodes"))
+```
+
+Below we visualize the cluster assignments using each parameter on a UMAP. 
+This isn't particularly useful because of clusters, but it can be helpful to see any values of K that may have obvious over or under clustering. 
+
+```{r, fig.height=15, fig.width=10}
+# look at clustering results across all params
+ggplot(umap_df, aes(x = UMAP1, y = UMAP2, color = cluster)) +
+  geom_point(alpha = 0.5, size = 0.1) +
+  facet_grid(
+    cols = vars(library_id),
+    rows = vars(nn_param)
+  ) +
+  theme(aspect.ratio = 1)
+```
+
+### Clustering statistics 
+
+Below we calculate a series of statistics: 
+
+- Average silhouette width: This metric evaluates cluster separation. 
+Cells with large positive silhouette widths are closer to other cells in the same cluster than to cells in different clusters. 
+- Average cluster purity: This metric also evaluates cluster separation and tells us the proportion of neighboring cells that are assigned to the same cluster. 
+Higher purity values indicate that clusters are well separated. 
+- Cluster stability: This evaluates how stable the clustering is to input data. 
+Higher values of cluster stability indicate more reproducible clusters. 
+
+```{r, warning=FALSE}
+# get a combined stats dataframe with purity and width for all clusters
+all_stats_df <- sce_list |>
+  purrr::imap(\(sce, library_id){
+    get_cluster_stats(sce, library_id, all_cluster_results)
+  }) |>
+  dplyr::bind_rows(.id = "library_id")
+```
+
+```{r}
+# summarize the silhouette width and purity for each nn_param used
+summary_stats_df <- all_stats_df |>
+  dplyr::group_by(library_id, nn_param) |>
+  dplyr::summarise(
+    mean_width = mean(width),
+    mean_purity = mean(purity)
+  ) |>
```

Comment:
I'd like to see some error lines in these plots too to assess overlap, if we're plotting means (that said, did you try maybe a `sina` with a red mean dot like the downstream analysis clustering report?). On one hand I'd say toss an a standard error calculation here too, but I actually you can probably skip a lot (all?) of this wrangling using `stat_summary()` which should plot your mean & SE automagically.